triton_kernel_wrap shouldn't access FakeTensor.data_ptr

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122309
* __->__ #122418

The comment suggests that we need to replace all FakeTensors with real
tensors. `torch.empty` doesn't actually return a real Tensor because
FakeTensorMode is active!

We disable torch dispatch so that torch.empty actually returns a real Tensor.

The motivation for this PR is that we're trying to ban
FakeTensor.data_ptr (or at least warn on it) in torch.compile. See the
next PR up in the stack

Test Plan:
- Existing tests.

zou3519 committed:
triton_kernel_wrap shouldn't access FakeTensor.data_ptr

The comment suggests that we need to replace all FakeTensors with real
tensors. `torch.empty` doesn't actually return a real Tensor because
FakeTensorMode is active!

We disable torch dispatch so that torch.empty actually returns a real Tensor.

The motivation for this PR is that we're trying to ban
FakeTensor.data_ptr (or at least warn on it) in torch.compile. See the
next PR up in the stack

Test Plan:
- Existing tests.

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -140,7 +140,8 @@ def generate_ttir(kernel, kwargs):
         elif isinstance(a, FakeTensor):
-            ordered_args[name] = torch.empty(2, dtype=a.dtype)
+            with torch._C._DisableTorchDispatch():
+                ordered_args[name] = torch.empty(2, dtype=a.dtype, device=a.device)
```
I don't know if we actually need to return a CUDA Tensor here. I can try on a CPU Tensor and see what happens if matters

oulgen replied:
My previous code did not return a cuda tensor, so perhaps not needed. After all, this code will not get executed it is only to generate metadata at triton side.

zou3519 replied:
Okay, let's try CPU

zou3519 committed:
Update on "triton_kernel_wrap shouldn't access FakeTensor.data_ptr"


The comment suggests that we need to replace all FakeTensors with real
tensors. `torch.empty` doesn't actually return a real Tensor because
FakeTensorMode is active!

We disable torch dispatch so that torch.empty actually returns a real Tensor.

The motivation for this PR is that we're trying to ban
FakeTensor.data_ptr (or at least warn on it) in torch.compile. See the
next PR up in the stack

Test Plan:
- Existing tests.

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
